docs/FoLiA-langcat.1
.TH FoLiA-langcat 1 "2020 jan 19"

.SH NAME
FoLiA-langcat - a FoLiA language detector
.SH SYNOPSIS

FoLiA-langcat [options] FILE

FoLiA-langcat [options] DIR

.SH DESCRIPTION
.B FoLiA-lancat
processes text nodes in FoLiA documents and assigns a language attribute to
each <t> node.
It uses Language Models as provided by Gertjan van Noord's language guesser TextCat.

This program is called 'textcat' 'libtextcat' or 'libexttextcat'
(depending on your configuration.) and is installed in /usr/share/libexttextcat/
or the like.
See also (http://odur.let.rug.nl/~vannoord/TextCat/)

When a DIR is provided,
.B FoLiA-langcat
will process all FoLiA files in DIR and for each file.xml it will produce
a file with a name file.lc.xml

When a FILE is provided,
.B FoLiA-langcat
will simply process that file and create file.lc.xml


.SH OPTIONS
.B --config
config
.RS
read the configuration for the 'textcat' LM from this file.

A typical configuration file might be:
.nf
#
# NB We use the iso639-2 codes here.
#
/usr/share/libexttextcat/nl.lm	dut
/usr/share/libexttextcat/de.lm	ger
.fi

This will detect Dutch and German, and assigns attributes 'dut' or 'ger'
accordingly.

.RE

.B --lang
lan
.RS
for unindentified text, assign the attribute 'lan'. (Default is 'dut')
.RE

.B --all
.RS
assign ALL detected language categories. The default is to assign the most
probable one.
.RE

.B --class
classname
.RS
The FoLiA classname of the <str> nodes that are created. (default OCR).
.RE

.B --tags
tag1,tag2,...,tagn
.RS
Process text in
.B
all
nodes of type <tag1>...<tagn>.
The default is to process text from <p> nodes.
.RE

.B -s
.RS
process FoLiA <str> nodes. OBSOLETE. Shorthand for --tags=s
.RE

.B -V
or
.B --version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
get help
.RE

.B -v
.RS
be more verbose
.RE

.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot

Martin Reynaert

e\-mail: lamasoftware@science.ru.nl


NEWS
foliautils 0.16  07-01-2021
[Ko vd Sloot]
* requires libfolia 2.7 or above
* provenance data is better for a lot of modules
* added better checking on invalid NCnames in some modules.
* FoLiA-abby:
  - a lot of refactoring and additions to handle font/style information
* FoLiA-pm:
  - Notes are handled correctly now
  - fixed error in xlink attributes
* FoLiA-page:
  - more types of Page files are handled now
  - fixed annotation declarations
  - fixed offset calculation (due to change in FoLiA's opinion on those)
  - page number is added as a <br> node and in the metadata
  - added a --trusttokens option. This means that Word items in the Page file
    are added as Word's in the FoLiA, embedded in Sentences.
  - added a --norefs option to avoid adding references to the original texts
* FoLiA-correct:
  - make sure that the default is to run on 1 thread
  - added a --rebase-inputclass option

* FoLiA-alto:
  - the -t option was not always handled correctly

[Maarten van Gompel]
* FoLiA-benchmark: guard against compiler optimisation #48

foliautils 0.15  15-09-2020
[Maarten van Gompel]
* FoLiA-txt: check if a string is empty after normalisation (fix for #46)
[Ko vd Sloot]
* folia-correct: fix one-off error in hemp handling (when no hemp was found) #45
* some refactoring
  - centralized definition of XML_PARSER_OPTIONS
* bugfix in threading

foliautils 0.14  15-04-2020
[Martin Reynaert]
* updated man pages

[Ko vd Sloot]
* added man pages
* revised usage() in many modules
* the default seperator in FoLiA-stats is '_' now
* fix for: https://github.com/LanguageMachines/foliautils/issues/37
* fix for: https://github.com/LanguageMachines/foliautils/issues/41
* adapted to changes in libfolia
* many small code refactorings
* FoLiA-correct is improved a lot, allowing ngram corrections in FoLiA
* FoLiA-stats accepts a 'word_in_doc' mode now
* FoLiA-alto by default ceated <w> nodes now. use --oldstring to get <str>
* improved a lot in tests/
* many small fixes

foliautils 0.13  30-08-2019
[Ko vd Sloot]
bug fixes:
* fix for https://github.com/LanguageMachines/foliautils/issues/35
* fix for https://github.com/LanguageMachines/foliautils/issues/36

[Maarten van Gompel]
new features:
* FoLiA-wordtranslate.cxx: frog should be able to deal with
  spaces now, no need for ugly non-breaking space hack that now causes
  other problems in frog (LanguageMachines/frog#77)
* FoLiA-wordtranslate.cxx: adding ability to constrain by language


foliautils 0.12  29-05-2019
[Ko vd Sloot]
Release for FoLiA 2.0.

foliautils 0.11  15-05-2019
[Ko vd Sloot]
Stabilizing release for pre FoLiA 2.0

* Updated and added some tests
* started moving common code to a separate file and build a library
  (libfoliautils)
  - hemp detection is one of them

* FoLiA-stats:
  - added possibility to read a list of directories + file-names to process
    into seperate output directories. (could be generalized to other programs)
  - better hemp detection
* FoLiA-correct:
  - use same hempo detections as FoLiA-stats
* FoLiA-abby:
  - support more flavors
* FoLiA-clean:
  - avoid removing the last remaining tekt on <w> nodes
  - cleaning of tokenization now works


foliautils 0.10  29-11-2018
[Ko vd Sloot]
  * fixed icu:namespace issues
  * added FoLiA-abby, an ABBY to FoLiA convertor
  * src/FoLiA-abby.cxx, src/FoLiA-page.cxx, src/FoLiA-pm.cxx:
     - Allow 'none' value for --prefix
  * src/FoLiA-page.cxx, src/FoLiA-hocr.cxx: fixed Alignment info
  * src/FoLiA-correct.cxx:
     - fixed a problem with correction of the last word of a trigram.
     - fix correction of paragraphs with only deeper text
     - The --rank option accepts more flavors of files
  * src/FoLiA-stats.cxx:
     - added a --detokenize option
  * several minor fixes, refactorings etc.
  * updated tests

foliautils 0.9.2  05-06-2018
[Ko vd Sloot]
Bug fix release:
 * append small prefixes to output filenames, to ALWAYS avoid names starting with
   a numeric value.
   'FPM-' for FoLiA-pm. 'FP-' for FoLiA-page, 'FH-' for FoLiA-hocr
   Can bet set witth --prefix
 * FoLiA-stats.cxx:
   - added --collect to usage() and 'man' page
 * FoLiA-correct:
   - added --inputclass and --outputclass parameters (must be different)
   - Don't crash on empty text.

foliautils 0.9.1  17-05-2018
[Ko vd Sloot]
Bug fix release: the tests directory wasn't included in the distribution

foliautils 0.9  16-05-2018
[Ko vd Sloot]
* FoLiA-stats.cxx:
  - added a --collect option, to create files with all n-grams
    together
  - clearer message in FoLiA-stats when no results were found
  - extract text from deeper nodes, if needed
  - fixed out-of-bounds problem
  - now fails when every inputfile fails
* FoLiA-txt:
  - now fails when every inputfile fails
* avoid xml:id's starting with a number. Add "id-" in front.
* added more tests

[Maarten van Gompel]
* added codemeta.json


foliautils 0.8  19-02-2018
[Ko vd Sloot]
* added -R option to FoLiA-collect
* FoLiA-collect now can work in parallel (-t option)
* modernized configuration, whit better Max OSX support (including OpenMP)
* all modules end with an exit code now.
* added more tests to 'make check'
* added output of Type-Token Ratio's (also in degrees)
* several bugfixes.
* code cleanup and refactoring, some speedup too

foliautils 0.7  24-10-2017
[ko vd Sloot]
 * updated and expanded tests
 * fixed offset calculations in FoLiA-hocr, FoLiA-page.cxx
   and FoLiA-alto. We use unicode points now. (needed for folia v1.5 and above)
 * Changed 'modes' in FoLiA-stats, to be a bit more comprehensible
 * fixed problem with metadatatype when 'foreign-data' is present
 * enhanced FoLiA-clean. Still not done...
 * switched to dynamic OMP scheduling in most programs.
   (which process files with probably big differences in processing time)
 * small bugfixes.
 * general cleanup and refactoring

[Maarten van Gompel]
 * Added and improved FoLiA-wordtranslate.cxx

foliautils 0.6  04-04-2017
This is an intermediate release!!

src/FoLiA-langcat.cxx
/*
  Copyright (c) 2014 - 2021
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of foliautils

  foliautils is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliautils is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/foliautils/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include <dirent.h>
#include <unistd.h>
#include <cstdlib>
#include <string>
#include <cstring>
#include <vector>
#include <algorithm>
#include <stdexcept>
#include <iostream>
#include <fstream>
#include "ticcutils/StringOps.h"
#include "libfolia/folia.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/CommandLine.h"
#include "foliautils/common_code.h"
#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif
#include "ucto/my_textcat.h"

using namespace	std;
using namespace	folia;

const string ISO_SET = "http://raw.github.com/proycon/folia/master/setdefinitions/iso639_3.foliaset";

bool verbose = false;

void setlang( FoliaElement* e, const string& lan ){
  // append a LangAnnotation child of class 'lan'
  vector<LangAnnotation*> lav;
  try {
    lav = e->annotations<LangAnnotation>(ISO_SET);
  }
  catch (...){
  }
  if ( lav.empty() ){
    KWargs args;
    args["class"] = lan;
    args["set"] = ISO_SET;
    LangAnnotation *node = new LangAnnotation( args, e->doc() );
    e->append( node );
  }
  else {
    bool present = false;
    for ( const auto it : lav ){
      if ( it->cls() == lan ){
	present = true;
	break;
      }
    }
    if ( !present ){
      KWargs args;
      args["class"] = lan;
      args["set"] = ISO_SET;
      LangAnnotation *node = new LangAnnotation( args, e->doc() );
      Alternative *a = new Alternative( );
      a->append( node );
      e->append( a );
    }
  }
}

void addLang( const TextContent *t,
	      const vector<string>& lv,
	      bool doAll ){
  //
  // we expect something like [dutch][french]
  //
  string val;
  for ( const auto& l : lv ){
    setlang( t->parent(), l );
    if ( !doAll )
      break;
  }
}

vector<FoliaElement*> gather_nodes( Document *doc, const string& docName,
				    const set<string>& tags ){
  vector<FoliaElement*> result;
  for ( const auto& tag : tags ){
    ElementType et;
    try {
      et = TiCC::stringTo<ElementType>( tag );
    }
    catch ( ... ){
#pragma omp critical (logging)
      {
	cerr << "the string '" << tag
	     << "' doesn't represent a known FoLiA tag" << endl;
	exit(EXIT_FAILURE);
      }
    }
    vector<FoliaElement*> v = doc->doc()->select( et );
#pragma omp critical (logging)
    {
      cout << "document '" << docName << "' has " << v.size() << " "
	   << tag << " nodes " << endl;
    }
    result.insert( result.end(), v.begin(), v.end() );
  }
  return result;
}


void procesFile( const TextCat& tc,
		 const string& outDir, const string& docName,
		 const string& default_lang,
		 const set<string>& tags,
		 bool doAll,
		 const string& cls,
		 const string& command ){
#pragma omp critical (logging)
  {
    cout << "process " << docName << endl;
  }
  Document *doc = 0;
  try {
    doc = new Document( "file='" + docName + "'" );
  }
  catch (const exception& e){
#pragma omp critical (logging)
    {
      cerr << "no document: " << e.what() << endl;
    }
    return;
  }
  doc->set_metadata( "language", default_lang );
  processor *proc = add_provenance( *doc, "FoLiA-langcat", command );
  if ( !doc->declared(  AnnotationType::LANG, ISO_SET ) ){
    KWargs args;
    args["processor"] = proc->id();
    doc->declare( AnnotationType::LANG, ISO_SET, args );
  }
  string outName;
  if ( !outDir.empty() ){
    outName = outDir + "/";
  }
  string::size_type pos = docName.rfind("/");
  if ( pos != string::npos ){
    outName += docName.substr( pos+1);
  }
  else {
    outName += docName;
  }
  string::size_type xml_pos = outName.find(".folia.xml");
  if ( xml_pos == string::npos ){
    xml_pos = outName.find(".xml");
  }
  outName.insert( xml_pos, ".lang" );
  if ( !TiCC::createPath( outName ) ){
#pragma omp critical (logging)
    {
      cerr << "unable to open output file " << outName << endl;
      cerr << "does the outputdir exist? And is it writabe?" << endl;
    }
    exit( EXIT_FAILURE);
  }
  vector<FoliaElement*> nodes = gather_nodes( doc, docName, tags );
  for ( const auto& node : nodes ){
    const TextContent *t = 0;
    string id = "NO ID";
    try {
      id = node->id();
      t = node->text_content(cls);
    }
    catch (...){
    }
    if ( t ){
      string para = t->str(cls);
      para = TiCC::trim( para );
      if ( para.empty() ){
	if ( verbose )
#pragma omp critical (logging)
	  {
	    cerr << "WARNING: no textcontent " << id << endl;
	  }
      }
      else {
	TiCC::to_lower( para );
	vector<string> lv = tc.get_languages( para );
	addLang( t, lv, doAll );
      }
    }
    else if ( verbose ){
#pragma omp critical (logging)
      {
	cerr << "WARNING: no textcontent " << id << endl;
      }
    }
  }
  doc->save(outName);
  delete doc;
}


void usage(){
  cerr << "Usage: [options] dir/filename " << endl;
  cerr << "  add language information to a file or files in FoLiA XML format" << endl;
  cerr << "  The files must have extension '.folia.xml' or '.xml'" << endl;
  cerr << "  or their .gz or .bz2 variants" << endl;
  cerr << "\t--config=<file>\t use LM config from 'file'" << endl;
  cerr << "\t--lang=<lan>\t use 'lan' for unindentified text. (default 'nld')" << endl;
  cerr << "\t--tags=t1,t2,..\t examine text in all <t1>, <t2> ...  nodes. (default is to use the <p> nodes)." << endl;
  cerr << "\t-s\t\t (obsolete) shorthand for --tags='str'" << endl;
  cerr << "\t--all\t\t assign ALL detected languages to the result. (default is to assign the most probable)." << endl;
  cerr << "\t--class=<cls>\t use 'cls' as the FoLiA classname for searching text. "
       << endl;
  cerr << "\t\t\t (default 'OCR')" << endl;
  cerr << "\t-t <threads>\n\t--threads <threads> Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-O path\t\t output path" << endl;
  cerr << "\t-V or --version\t show version info." << endl;
  cerr << "\t-v\t\t verbose" << endl;
  cerr << "\t-h or --help\t this messages." << endl;
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts( "svVhO:t:", "all,lang:,class:,config:,help,version,tags:,threads:" );
  try {
    opts.init( argc, argv );
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage();
    exit( EXIT_FAILURE );
  }
  if ( opts.extract( 'V' ) || opts.extract( "version" ) ){
    cout << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract( 'h' ) || opts.extract( "help") ){
    usage();
    exit(EXIT_SUCCESS);
  }
  string outDir;
  string config = "./config/tc.txt";
  string lang = "nld";
  string cls = "OCR";
  verbose = opts.extract( 'v' );
  string value;
  if ( opts.extract( 't', value )
       || opts.extract( "threads", value ) ){
#ifdef HAVE_OPENMP
    int numThreads = 1;
    if ( TiCC::lowercase(value) == "max" ){
      numThreads = omp_get_max_threads() - 2;
    }
    else if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
    omp_set_num_threads( numThreads );
#else
    cerr << "FoLiA-langcat: OpenMP support is missing. -t option is not supported" << endl;
    exit( EXIT_FAILURE );
#endif
  }
  bool doAll = opts.extract( "all" );
  bool doStrings = opts.extract( 's' );
  opts.extract( "config", config );
  opts.extract( "lang", lang );
  opts.extract( "class", cls );
  opts.extract( 'O', outDir );
  set<string> tags;
  string tagsstring;
  opts.extract( "tags", tagsstring );
  if ( !tagsstring.empty() ){
    vector<string> parts = TiCC::split_at( tagsstring, "," );
    for( const auto& t : parts ){
      tags.insert( t );
    }
  }
  if ( doStrings ){
    if ( !tags.empty() ){
      cerr << "--tags and -s conflict." << endl;
      exit(EXIT_FAILURE);
    }
    else {
      tags.insert( "str" );
    }
  }
  if ( tags.empty() ){
    tags.insert( "p" );
  }
  if ( !opts.empty() ){
    cerr << "unsupported options : " << opts.toString() << endl;
    usage();
    exit(EXIT_FAILURE);
  }
  string command = "FoLiA-langcat " + opts.toString();
  vector<string> fileNames = opts.getMassOpts();
  if ( fileNames.empty() ){
    cerr << "missing input file(s)" << endl;
    exit( EXIT_FAILURE );
  }
  else if ( fileNames.size() == 1 ){
    string name = fileNames[0];
    try {
      fileNames = TiCC::searchFilesMatch( name, "*.xml*", false );
    }
    catch ( ... ){
      cerr << "no matching xml file found: '" << name << "'" << endl;
      exit( EXIT_FAILURE );
    }
  }
  TextCat TC( config );
  if ( !TC.isInit() ){
    cerr << "unable to init from: " << config << endl;
    exit(EXIT_FAILURE);
  }

  size_t toDo = fileNames.size();
  if ( toDo > 1 ){
    cout << "start processing of " << toDo << " files " << endl;
  }
#pragma omp parallel for firstprivate(TC) shared(fileNames,toDo) schedule(dynamic)
  for ( size_t fn=0; fn < toDo; ++fn ){
    string docName = fileNames[fn];
    procesFile( TC, outDir, docName, lang, tags, doAll, cls, command );
  }
  return EXIT_SUCCESS;
}
